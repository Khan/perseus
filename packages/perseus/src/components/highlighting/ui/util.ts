/**
 * Utility functions for highlighting UI.
 */
import {rangesOverlap, intersectRanges} from "../ranges";

import type {DOMRange, Position, Rect} from "./types";

/**
 * Given two positions relative to the same origin, return `child`'s position
 * relative to `parent`'s position.
 */
export function getRelativePosition(
    child: Position,
    parent: Position,
): Position {
    return {
        left: child.left - parent.left,
        top: child.top - parent.top,
    };
}

/**
 * `getRelativePosition`, but for Rects.
 *
 * Given a rectangle and a position relative to the same origin, return a new
 * rectangle of the same size, whose position is relative to the given
 * position.
 */
export function getRelativeRect(
    {left, top, width, height}: Rect,
    parent: Position,
): Rect {
    return {
        ...getRelativePosition({left, top}, parent),
        width,
        height,
    };
}

/**
 * Given a DOMRange, return a list of client-relative rects that cover the text
 * inside the DOMRange.
 *
 * This behaves differently than the browser's built-in `getClientRects` method
 * for DOMRanges, because this method only gets rects for the *text* nodes,
 * whereas the browser's method will also return rects for any fully-contained
 * *element* nodes.
 *
 * For example, if an entire paragraph is selected, then `getClientRects` will
 * return a rectangle that covers the entire paragraph block, whereas this
 * method will only return rectangles for the text inside the paragraph.
 */
export function getClientRectsForTextInRange(
    range: DOMRange,
): ReadonlyArray<Rect> {
    const mutableRects = [];
    addClientRectsForTextInNodeAndRange(
        range.commonAncestorContainer,
        range,
        mutableRects,
    );
    return mutableRects;
}

/**
 * A helper method for getClientRectsForTextInRange.
 *
 * Given a node and a range, find text that is contained within both the node
 * and the range by recursing down the node's subtree, and push the text's
 * client rects onto the given `mutableRects` array.
 */
function addClientRectsForTextInNodeAndRange(
    node: Node,
    range: DOMRange,
    mutableRects: Array<Rect>,
): void {
    const nodeContentsRange = new Range();
    nodeContentsRange.selectNodeContents(node);

    if (node.nodeType === Node.TEXT_NODE) {
        // Get the text that appears both in this text node and in the range
        // we're searching.
        const intersectedRange = intersectRanges(range, nodeContentsRange);
        if (intersectedRange) {
            addClientRectsForText(node, intersectedRange, mutableRects);
        }
    } else if (node.nodeType === Node.ELEMENT_NODE) {
        // Don't bother deep-traversing a node that doesn't even overlap the
        // range we're searching.
        if (!rangesOverlap(range, nodeContentsRange)) {
            return;
        }

        // NOTE(mdr): Safari returns a non-iterable `childNodes` object, so
        //     convert it to an array.
        for (const child of Array.from(node.childNodes)) {
            // TODO(mdr): If this parent has overflow: hidden, bound the child
            //     rects accordingly. This would allow us to remove the hack
            //     we added to .perseus-sr-only.
            addClientRectsForTextInNodeAndRange(child, range, mutableRects);
        }
    }
}

/**
 * Given a DOMRange of text fully contained by the given `textNode`, compute
 * rectangles that visually cover the range, and push them onto the given
 * `mutableRects` array.
 *
 * This function adjusts the rectangles to the line height of the text, to
 * match the visual behavior of native text selection, and remove annoying gaps
 * between lines that disrupt hover behavior.
 */
function addClientRectsForText(
    textNode: Node,
    textRange: DOMRange,
    mutableRects: Array<Rect>,
): void {
    const parentElement = textNode.parentElement;
    // @ts-expect-error - TS2345 - Argument of type 'HTMLElement | null' is not assignable to parameter of type 'Element'.
    const computedStyle = window.getComputedStyle(parentElement);

    // NOTE(mdr): I'm not sure how computed line height works in all contexts
    //     in all browsers. It's valid to specify line height as a ratio,
    //     relative to font size. Latest Chrome seems to transform this into a
    //     px value, but maybe some browsers don't, so I'm being defensive here
    //     and only using the computed line height if it's explicitly a px
    //     value. We could also handle other browsers by computing font size
    //     and doing the math ourselves, but let's punt that until we need it.
    let lineHeight = null;
    if (
        typeof computedStyle.lineHeight === "string" &&
        computedStyle.lineHeight.endsWith("px")
    ) {
        const parsedLineHeight = parseFloat(computedStyle.lineHeight);
        if (!isNaN(parsedLineHeight)) {
            // @ts-expect-error - TS2322 - Type 'number' is not assignable to type 'null'.
            lineHeight = parsedLineHeight;
        }
    }

    // NOTE(mdr): Safari returns a non-iterable `getClientRects` object, so
    //     convert it to an array.
    const boundingRects = Array.from(textRange.getClientRects());

    for (const boundingRect of boundingRects) {
        const rect: Rect = {
            left: boundingRect.left,
            top: boundingRect.top,
            height: boundingRect.height,
            width: boundingRect.width,
        };

        // Adjust the rectangle according to the text's line height. It should
        // keep the same bottom coordinate, so the top coordinate will move
        // accordingly.
        if (lineHeight !== null) {
            const heightIncrease = lineHeight - rect.height;
            rect.top -= heightIncrease;
            rect.height = lineHeight;
        }

        mutableRects.push(rect);
    }
}
