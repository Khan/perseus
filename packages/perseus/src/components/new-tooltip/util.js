// @flow
/**
 * Utility functions for the NewTooltip component.
 */
import Color from "@khanacademy/wonder-blocks-color";

import type {NewTooltipProps} from "./types.js";

/**
 * Given a set of NewTooltipProps, return the `backgroundColor` and `textColor`
 * of the tooltip. Used in TooltipArrow and TooltipBubble.
 */
export function getTooltipColors(tooltipProps: {
    color: NewTooltipProps["color"],
    inverted: NewTooltipProps["inverted"],
    ...
}): {|
    backgroundColor: string,
    textColor: string,
|} {
    const {color, inverted} = tooltipProps;

    if (inverted) {
        return {backgroundColor: color, textColor: Color.white};
    }
    return {backgroundColor: Color.white, textColor: color};
}
