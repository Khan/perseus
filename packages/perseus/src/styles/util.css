.perseus-sr-only {
    border: 0;
    clip: rect(0, 0, 0, 0);
    font-size: 0;
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
}
.perseus-clearfix:before,
.perseus-clearfix:after {
    content: "";
    display: table;
}
.perseus-clearfix:after {
    clear: both;
}
