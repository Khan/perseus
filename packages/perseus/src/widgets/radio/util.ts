import * as i18n from "@khanacademy/wonder-blocks-i18n";

/**
 * Given a choice's position in the radio widget, return the corresponding
 * "choice letter". (For example, in English, `getChoiceLetter(0)` is "A",
 * `getChoiceLetter(1)` is "B", etc.)
 */
export function getChoiceLetter(pos: number): string {
    /* I18N: This is a list of single-character labels that will appear in
     * front of multiple-choice options. For instance, a multiple-choice
     * question with three options would display
     *  (A) first option
     *  (B) second option
     *  (C) third option
     * There must be spaces between each of the different characters. The
     * characters will show up next to options in the order that they are
     * listed here. Most multiple choice questions have 5 or fewer options.
     */
    const lettersString = i18n._(
        "A B C D E F G H I J K L M N O P Q R S T U V W X Y Z",
    );

    const letters = lettersString.split(" ");

    if (pos < letters.length) {
        // If the position we need is listed in the localized string, use that.
        return letters[pos];
    }
    // If we're out of letters, give up and return a space.
    return " ";
}
