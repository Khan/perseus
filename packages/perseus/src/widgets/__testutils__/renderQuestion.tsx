import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
