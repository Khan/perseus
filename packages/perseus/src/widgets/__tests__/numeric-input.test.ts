import {screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {errors} from "../../util/answer-types";
import {
    question1AndAnswer,
    multipleAnswers,
    percentageProblem,
    multipleAnswersWithDecimals,
    question1,
    duplicatedAnswers,
    withCoefficient,
} from "../__testdata__/numeric-input.testdata";
import {
    maybeParsePercentInput,
    NumericInput,
    unionAnswerForms,
} from "../numeric-input";

import {renderQuestion} from "./renderQuestion";

import type {Rubric} from "../numeric-input";

describe("numeric-input widget", () => {
    const [question, correct, incorrect] = question1AndAnswer;

    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("Should accept the right answer", () => {
        // Arrange
        const {renderer} = renderQuestion(question);

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), correct);

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("Should render predictably", () => {
        // Arrange
        const {container} = renderQuestion(question);
        expect(container).toMatchSnapshot("first render");

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), correct);

        // Assert
        expect(container).toMatchSnapshot("after interaction");
    });

    it("should reject an incorrect answer", () => {
        // Arrange
        const {renderer} = renderQuestion(question);

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), incorrect);

        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });
});

describe("static function getOneCorrectAnswerFromRubric", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("can get one correct answer from a rubric with multiple answers", () => {
        const widget = multipleAnswersWithDecimals.widgets["numeric-input 1"];
        const widgetOptions = widget && widget.options;
        const answers: ReadonlyArray<any> =
            (widgetOptions && widgetOptions.answers) || [];

        const singleAnswer = NumericInput.getOneCorrectAnswerFromRubric({
            answers,
            labelText: "",
            size: "medium",
            static: false,
            coefficient: false,
        });
        expect(singleAnswer).toBe("12.2");
    });

    it("can get one correct answer from a rubric with one answer", () => {
        const widget = question1.widgets["numeric-input 1"];
        const widgetOptions = widget && widget.options;
        const answers: ReadonlyArray<any> =
            (widgetOptions && widgetOptions.answers) || [];
        const singleAnswer = NumericInput.getOneCorrectAnswerFromRubric({
            answers,
            labelText: "",
            size: "medium",
            static: false,
            coefficient: false,
        });
        expect(singleAnswer).toBe("1252");
    });

    it("can not get a correct answer from a rubric with no answer", () => {
        const answers: Array<never> = [];
        const singleAnswer = NumericInput.getOneCorrectAnswerFromRubric({
            answers,
            labelText: "",
            size: "medium",
            static: false,
            coefficient: false,
        });
        expect(singleAnswer).toBeUndefined();
    });

    it("supports error bars", () => {
        const rubric: Rubric = {
            answers: [
                {
                    status: "correct",
                    value: 1.0,
                    maxError: 0.2,
                    answerForms: ["decimal"],
                    simplify: "",
                    strict: false,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };
        const singleAnswer = NumericInput.getOneCorrectAnswerFromRubric(rubric);
        expect(singleAnswer).toBe("1 Â± 0.2");
    });
});

describe("static function validate", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("with a simple value", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0,
                    simplify: "",
                    strict: false,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "1",
        } as const;

        const score = NumericInput.validate(useInput, rubric);

        expect(score).toMatchInlineSnapshot(`
            {
              "earned": 1,
              "message": null,
              "total": 1,
              "type": "points",
            }
        `);
    });

    it("with nonsense", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0,
                    simplify: "",
                    strict: false,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "sadasdfas",
        } as const;

        const score = NumericInput.validate(useInput, rubric);

        expect(score).toMatchInlineSnapshot(`
            {
              "message": "We could not understand your answer. Please check your answer for extra text or symbols.",
              "type": "invalid",
            }
        `);
    });

    // Don't default to validating the answer as a pi answer
    // if answerForm isn't set on the answer.
    // The answer value, userInput.currentValue, and
    // the omission of answerForms in the answer are
    // important to the test.
    // https://khanacademy.atlassian.net/browse/LC-691
    it("doesn't default to validating pi", () => {
        const rubric: Rubric = {
            answers: [
                {
                    maxError: null,
                    message: "",
                    simplify: "required",
                    status: "correct",
                    strict: true,
                    value: 45.289,
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: false,
        };

        const userInput = {
            // (pi / 12) * 173 = 45.291
            // within the 0.01 margin of error
            // to trigger the pi validation flow
            currentValue: "45.282",
        } as const;

        const score = NumericInput.validate(userInput, rubric);

        expect(score.message).not.toBe(errors.APPROXIMATED_PI_ERROR);
        expect(score.message?.includes("pi")).toBeFalsy();
    });

    it("still validates against pi if provided in answerForms", () => {
        const rubric: Rubric = {
            answers: [
                {
                    maxError: null,
                    message: "",
                    simplify: "required",
                    status: "correct",
                    strict: false,
                    value: 311.01767270538954,
                    answerForms: ["pi"],
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: false,
        };

        const userInput = {
            currentValue: "99 pi",
        } as const;

        const score = NumericInput.validate(userInput, rubric);

        expect(score).toMatchInlineSnapshot(`
        {
          "earned": 1,
          "message": null,
          "total": 1,
          "type": "points",
        }
    `);
    });

    it("with a strict answer", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0,
                    simplify: "",
                    strict: true,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "1.0",
        } as const;

        const score = NumericInput.validate(useInput, rubric);

        expect(score).toMatchInlineSnapshot(`
            {
              "earned": 1,
              "message": null,
              "total": 1,
              "type": "points",
            }
        `);
    });

    it("with a strict answer and max error is outside range", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0.2,
                    simplify: "",
                    strict: true,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "1.3",
        } as const;

        const score = NumericInput.validate(useInput, rubric);

        expect(score).toMatchInlineSnapshot(`
            {
              "earned": 0,
              "message": null,
              "total": 1,
              "type": "points",
            }
        `);
    });

    it("with a strict answer and max error is inside range", () => {
        const rubric: Rubric = {
            answers: [
                {
                    value: 1,
                    status: "correct",
                    maxError: 0.2,
                    simplify: "",
                    strict: true,
                    message: "",
                },
            ],
            labelText: "",
            size: "normal",
            static: false,
            coefficient: true,
        };

        const useInput = {
            currentValue: "1.12",
        } as const;

        const score = NumericInput.validate(useInput, rubric);

        expect(score).toMatchInlineSnapshot(`
            {
              "earned": 1,
              "message": null,
              "total": 1,
              "type": "points",
            }
        `);
    });
});

describe("Numeric input widget", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("can handle multiple correct answers (Part one)", () => {
        // Arrange
        const {renderer} = renderQuestion(multipleAnswers);

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), "1");

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can handle multiple correct answers (Part two)", () => {
        // Arrange
        const {renderer} = renderQuestion(multipleAnswers);

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), "2");

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can handle duplicated answers", () => {
        // Arrange
        const {renderer} = renderQuestion(duplicatedAnswers);

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), "2.4");

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can handle coefficients", () => {
        // Arrange
        const {renderer} = renderQuestion(withCoefficient);

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), "1.0");

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("handles answers that are percentages", () => {
        // Arrange
        const {renderer} = renderQuestion(percentageProblem);

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), "33%");

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("handles answers that are decimals", () => {
        // Arrange
        const {renderer} = renderQuestion(multipleAnswersWithDecimals);

        // Act
        userEvent.paste(screen.getByRole("textbox", {hidden: true}), "12.2");

        // Assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("styles differently on mobile", () => {
        const {container} = renderQuestion(multipleAnswersWithDecimals, {
            // I wish this was more clear but this is how mobile
            // rendering is triggered
            customKeypad: true,
        });

        expect(container).toMatchSnapshot("mobile render");
    });

    it("can be focused", () => {
        const {renderer} = renderQuestion(question1);

        // Act
        const gotFocus = renderer.focus();

        // Assert
        expect(gotFocus).toBeTrue();
        // eslint-disable-next-line testing-library/no-node-access
        expect(document.activeElement).toBe(
            screen.getByRole("textbox", {hidden: true}),
        );
    });

    it("calls the interaction callback on focus", () => {
        const testCallback = jest.fn();

        const {renderer} = renderQuestion(question1, {
            interactionCallback: testCallback,
        });

        // Act
        renderer.focus();

        // Assert
        expect(testCallback).toHaveBeenCalled();
    });

    it("can be blurred", () => {
        const {renderer} = renderQuestion(question1);

        // Act
        const input = screen.getByRole("textbox", {hidden: true});
        input.focus();
        renderer.blur();

        // Assert
        // eslint-disable-next-line testing-library/no-node-access
        expect(document.activeElement).not.toBe(
            screen.getByRole("textbox", {hidden: true}),
        );
    });
});

describe("unionAnswerForms utility function", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("removes duplicates", () => {
        // arrange
        const forms = [
            [
                {
                    simplify: "required",
                    name: "integer",
                },
            ],
            [
                {
                    simplify: "required",
                    name: "integer",
                },
            ],
        ];

        // act
        // @ts-expect-error - TS2345 - Argument of type '{ simplify: string; name: string; }[][]' is not assignable to parameter of type 'readonly (readonly PerseusNumericInputAnswerForm[])[]'.
        const result = unionAnswerForms(forms);

        // assert
        expect(result).toBeArrayOfSize(1);
    });
});

describe("maybeParsePercentInput utility function", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("returns a percent if the input is a percent", () => {
        const result = maybeParsePercentInput("3%", false);

        expect(result).toBe(3);
    });

    it("returns the input if its not valid", () => {
        const result = maybeParsePercentInput("asdasd%", false);

        expect(result).toBe("asdasd%");
    });

    it("normalizes values", () => {
        const result = maybeParsePercentInput("20%", true);

        expect(result).toBe(0.2);
    });
});
