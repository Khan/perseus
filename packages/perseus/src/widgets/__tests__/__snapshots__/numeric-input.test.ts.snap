// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Numeric input widget styles differently on mobile: mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <div
            aria-label="Math input box Tap with one or two fingers to open keyboard"
            class="initial_4qg14c-o_O-input_1n1c032"
            role="textbox"
          >
            <div
              class="keypad-input"
              tabindex="0"
            >
              <div
                class="mq-editable-field mq-math-mode"
                style="background-color: white; min-height: 44px; min-width: 64px; max-width: 128px; box-sizing: border-box; position: relative; border-style: solid; border-color: rgba(33,36,44,0.50); border-radius: 4px; color: rgb(33, 36, 44); border-width: 1px;"
              >
                <span
                  class="mq-textarea"
                >
                  <span />
                </span>
                <span
                  class="mq-root-block mq-empty"
                  mathquill-block-id="1"
                  style="padding: 10px 11px 8px 11px; font-size: 18pt;"
                />
              </div>
            </div>
          </div>
        </div>
         
      </div>
    </div>
  </div>
</div>
`;

exports[`numeric-input widget Should render predictably: after interaction 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <div>
            <span>
              <div>
                <input
                  aria-label="Your answer:"
                  autocapitalize="off"
                  autocomplete="off"
                  autocorrect="off"
                  class="input_1t4b01j-o_O-LabelMedium_1rew30o-o_O-default_1r2dwp3-o_O-focused_1z0d9ul-o_O-input_19vw1dn"
                  id="input-with-examples-bnVtZXJpYy1pbnB1dCAx"
                  type="text"
                  value="1252"
                />
              </div>
              <div
                style="position: relative; height: 0px; display: none;"
              >
                <div
                  class="tooltipContainer"
                  style="position: absolute; left: 0px;"
                >
                  <div
                    style="display: block; position: relative; visibility: visible; left: 0px; top: -1px; width: 12px; height: 11px; margin-top: -1px; margin-bottom: -2px; z-index: 10;"
                  >
                    <div
                      style="display: block; height: 0px; width: 0px; position: absolute; left: 0px; top: -1px; border-right: 12px solid transparent; border-bottom: 12px solid #ccc;"
                    />
                    <div
                      style="display: block; height: 0px; width: 0px; position: absolute; left: 1px; top: 1px; border-right: 10px solid transparent; border-bottom: 10px solid white;"
                    />
                  </div>
                  <div
                    class="perseus-formats-tooltip preview-measure"
                    style="position: relative; top: 0px; left: 0px; border: 1px solid #ccc; box-shadow: 0 1px 3px #ccc; z-index: 9;"
                  >
                    <div
                      id="input-with-examples-bnVtZXJpYy1pbnB1dCAx"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <ul>
                            <li>
                              <strong>
                                Your answer should be
                              </strong>
                            </li>
                            <li>
                              an integer, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  6
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              a 
                              <em>
                                simplified proper
                              </em>
                               fraction, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  3/5
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              a 
                              <em>
                                simplified improper
                              </em>
                               fraction, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  7/4
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              a mixed number, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  1\\ 3/4
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              an 
                              <em>
                                exact
                              </em>
                               decimal, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  0.75
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              a multiple of pi, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  12\\ \\text{pi}
                                </span>
                                <span />
                              </span>
                               or 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  2/3\\ \\text{pi}
                                </span>
                                <span />
                              </span>
                            </li>
                          </ul>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </span>
          </div>
        </div>
         
      </div>
    </div>
  </div>
</div>
`;

exports[`numeric-input widget Should render predictably: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            5008 \\div 4 =
          </span>
          <span />
        </span>
         
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <div>
            <span>
              <div>
                <input
                  aria-label="Your answer:"
                  autocapitalize="off"
                  autocomplete="off"
                  autocorrect="off"
                  class="input_1t4b01j-o_O-LabelMedium_1rew30o-o_O-default_1r2dwp3-o_O-input_19vw1dn"
                  id="input-with-examples-bnVtZXJpYy1pbnB1dCAx"
                  type="text"
                  value=""
                />
              </div>
              <div
                style="position: relative; height: 0px; display: none;"
              >
                <div
                  class="tooltipContainer"
                  style="position: absolute; left: 0px;"
                >
                  <div
                    style="display: block; position: relative; visibility: visible; left: 0px; top: -1px; width: 12px; height: 11px; margin-top: -1px; margin-bottom: -2px; z-index: 10;"
                  >
                    <div
                      style="display: block; height: 0px; width: 0px; position: absolute; left: 0px; top: -1px; border-right: 12px solid transparent; border-bottom: 12px solid #ccc;"
                    />
                    <div
                      style="display: block; height: 0px; width: 0px; position: absolute; left: 1px; top: 1px; border-right: 10px solid transparent; border-bottom: 10px solid white;"
                    />
                  </div>
                  <div
                    class="perseus-formats-tooltip preview-measure"
                    style="position: relative; top: 0px; left: 0px; border: 1px solid #ccc; box-shadow: 0 1px 3px #ccc; z-index: 9;"
                  >
                    <div
                      id="input-with-examples-bnVtZXJpYy1pbnB1dCAx"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <ul>
                            <li>
                              <strong>
                                Your answer should be
                              </strong>
                            </li>
                            <li>
                              an integer, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  6
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              a 
                              <em>
                                simplified proper
                              </em>
                               fraction, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  3/5
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              a 
                              <em>
                                simplified improper
                              </em>
                               fraction, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  7/4
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              a mixed number, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  1\\ 3/4
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              an 
                              <em>
                                exact
                              </em>
                               decimal, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  0.75
                                </span>
                                <span />
                              </span>
                            </li>
                            <li>
                              a multiple of pi, like 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  12\\ \\text{pi}
                                </span>
                                <span />
                              </span>
                               or 
                              <span
                                style="white-space: nowrap;"
                              >
                                <span />
                                <span
                                  class="mock-TeX"
                                >
                                  2/3\\ \\text{pi}
                                </span>
                                <span />
                              </span>
                            </li>
                          </ul>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </span>
          </div>
        </div>
         
      </div>
    </div>
  </div>
</div>
`;
